### PR TITLE
Convert documentation to ReStructured Text (RST).

### PR DESCRIPTION
**Pull request by lukesneeringer**
This commit adds `pypandoc` and converts docstrings to RST from Markdown.

### CODE REVIEW

File: gapic/templates/$namespace/$name_$version/types/_enum.py.j2 [hunk 1]
```diff
@@ -1,6 +1,5 @@
 class {{ enum.name }}(_.enum.IntEnum):
-    """{{ enum.meta.doc|wrap(width=72, indent=4) }}
-    """
+    """{{ enum.meta.doc|rst(width=72, indent=4) }}"""
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/templates/$namespace/$name_$version/types/_enum.py.j2]: The filter handles that.

File: gapic/utils/lines.py [hunk 2]
```diff
@@ -44,31 +44,50 @@ def wrap(text: str, width: int, *, offset: int = None, indent: int = 0) -> str:
 
     # If the offset is None, default it to the indent value.
     if offset is None:
-        offset = indent
+        offset = indent + 3
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/utils/lines.py]: ```len('"""')```

File: gapic/utils/lines.py [hunk 3]
```diff
@@ -44,31 +44,50 @@ def wrap(text: str, width: int, *, offset: int = None, indent: int = 0) -> str:
...
+    if not text:
+        return first.strip()
+
+    # Tokenize the rest of the text to try to preserve line breaks
+    # that semantically matter.
+    tokens = []
+    token = ''
+    for line in text.split('\n'):
+        token += line + '\n'
+        if len(line) < width * 0.75:
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/utils/lines.py]: This is more or less arbitrary.

Basically when I was looking at what `wrap` did, I noticed that a lot of times where were meaningful newlines in proto comments that were getting plowed over. Something like.

```protobuf
// Do a thing.
// This rpc does a thing and passes it to another thing and returns a thing.
```

This makes the word wrapping algorithm respect that line break as a "hard break".

File: gapic/utils/rst.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    # Sanity check: If the text block does not appear to have any formatting,
+    # do not convert it.
+    # (This makes code generation significantly faster; calling out to pandoc
+    # is by far the most expensive thing we do.)
+    if not re.search(r'[|*`_[\]]', text):
+        answer = wrap(text, width=width, indent=indent, offset=indent + 3)
+    else:
+        # Convert from CommonMark to ReStructured Text.
+        answer = pypandoc.convert_text(text, 'rst',
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/utils/rst.py]: I am _willing_ to do this but would prefer not to. The wrap algorithm is much (much) faster and I view it as an implementation detail that we could adjust later if we so desired.

File: gapic/generator/generator.py [hunk 5]
```diff
@@ -60,6 +60,7 @@ def __init__(self, api_schema: api.API, *,
         )
 
         # Add filters which templates require.
+        self._env.filters['rst'] = utils.rst
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/generator/generator.py]: Done, thanks!

File: gapic/utils/lines.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/utils/lines.py]: However, I did not intend to do that here; it is supposed to be in the `rst` function itself. Removed this and added documentation there.

File: gapic/templates/$namespace/$name_$version/types/_enum.py.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply gapic/templates/$namespace/$name_$version/types/_enum.py.j2]: From `rst.py`:

```python
    # Add a newline to the end of the document if any line breaks are
    # already present.
    #
    # This causes the closing """ to be on the subsequent line only when
    # appropriate.
    if '\n' in answer:
        answer += '\n' + ' ' * indent
```

File: gapic/utils/lines.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (landrito) [reply gapic/utils/lines.py]: This is more a less a nit but context in comments might be helpful I mean.

File: gapic/utils/rst.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (landrito) [reply gapic/utils/rst.py]: Gotcha lets stick with what we have.

File: gapic/templates/$namespace/$name_$version/types/_enum.py.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (landrito) [reply gapic/templates/$namespace/$name_$version/types/_enum.py.j2]: Nit: Maybe we make it consistent across all jinja templates to do these docstrings on a single line.

[the remark this answers is not in the document]
Comment (landrito) [reply gapic/templates/$namespace/$name_$version/types/_enum.py.j2]: Ah I just reread the jinja templates. I thought these single line comments were inconsistent.

**Pull request merged**